build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/testfixtures/TestFixturesPlugin.java

```diff
@@ -109,8 +110,7 @@ public void execute(Task task) {
             ComposeExtension composeExtension = project.getExtensions().getByType(ComposeExtension.class);
             composeExtension.getUseComposeFiles().addAll(Collections.singletonList(DOCKER_COMPOSE_YML));
             composeExtension.getRemoveContainers().set(true);
-            composeExtension.getExecutable()
-                .set(project.file("/usr/local/bin/docker-compose").exists() ? "/usr/local/bin/docker-compose" : "/usr/bin/docker-compose");
+            composeExtension.getExecutable().set(dockerSupport.get().getDockerAvailability().dockerComposePath());
```

Comment:
I think this will throw an error if `dockerComposePath()` returns `null`. Additionally `getDockerAvailability()` is a relatively expensive operation, so we want to defer this unless we are actually running a build that will execute docker tasks. Thanksfully `getExecutable()` is a `Provider` so we can lazily evaluate this.

You can create a provider by injecting a `ProviderFactory` (like `FileSystemOperations` above) and then doing something like `getExecutable().set(providerFactory.provider(() -> dockersupport....))`. We'll need to deal with missing values too I suspect and provide a default convention value.

Reply:
The `apply` method calls `maybeSkipTask` up to 9 times, and each call invokes `getDockerAvailability`. To address this concern, I pulled `getDockerAvailability` out of `maybeSkipTask`. It is only calls once now, and values for isComposeAvailable() and composePath() are put in variables and used locally.

The old getExecutable().set() used "/usr/bin/docker" as a fallback. I re-added that as a fallback if composePath() is null. That should avoid an NPE and work in a bwc manner. I assume there is a better way to do this, but I am just looking to get this working with the new Docker Compose path. If you think additional cleanup is needed, let me know what should be done in this PR, or if it can be left to a follow up PR.

Reply:
> The apply method calls maybeSkipTask up to 9 times, and each call invokes getDockerAvailability. To address this concern, I pulled getDockerAvailability out of maybeSkipTask, so it is only called once now, and values for isComposeAvailable() and composePath() are put in variables for reuse in different places if necessary.

This isn't necessary. The `getDockerAvailablity` method is memoized and simply returns the previously computed result on subsequent calls. There's no need to do any refactoring to minimize the number of consecutive calls here.

Reply:
> The old getExecutable().set() used "/usr/bin/docker" as a fallback. I re-added that as a fallback if composePath() is null. That should avoid an NPE and work in a bwc manner. I assume there is a better way to do this, but I am just looking to get this working with the new Docker Compose path. If you think additional cleanup is needed, let me know what should be done in this PR, or if it can be left to a follow up PR.

This doesn't address my main concern though, which is that we don't want to reference `getDockerAvailability` at configuration time. We want to defer this to task execution time, which is why we want to wrap it in a Gradle `Provider`.

Reply:
@elasticmachine update branch

Reply:
@mark-vieira I reverted and used a Provider. Let me know if this is what you meant, or if you meant something else.